tests/filecheck/transforms/lift-arith-to-linalg.mlir

```diff
@@ -1,15 +1,64 @@
 // RUN: xdsl-opt %s -p lift-arith-to-linalg | filecheck %s
 
 builtin.module {
+// CHECK: builtin.module {
+
+  func.func @a() {
     %t0, %t1, %t2, %t3 = "test.op"() : () -> (tensor<8xf32>, tensor<8xf32>, tensor<8xf32>, tensor<8xf32>)
     %0 = arith.addf %t0, %t1 : tensor<8xf32>
     %1 = arith.subf %0, %t2 : tensor<8xf32>
     %2 = arith.mulf %1, %t3 : tensor<8xf32>
-}
+    func.return
+  }
 
-// CHECK-NEXT: builtin.module {
+// CHECK-NEXT: func.func @a() {
 // CHECK-NEXT:   %t0, %t1, %t2, %t3 = "test.op"() : () -> (tensor<8xf32>, tensor<8xf32>, tensor<8xf32>, tensor<8xf32>)
 // CHECK-NEXT:   %0 = linalg.add ins(%t0, %t1 : tensor<8xf32>, tensor<8xf32>) outs(%t0 : tensor<8xf32>) -> tensor<8xf32>
 // CHECK-NEXT:   %1 = linalg.sub ins(%0, %t2 : tensor<8xf32>, tensor<8xf32>) outs(%0 : tensor<8xf32>) -> tensor<8xf32>
 // CHECK-NEXT:   %2 = linalg.mul ins(%1, %t3 : tensor<8xf32>, tensor<8xf32>) outs(%1 : tensor<8xf32>) -> tensor<8xf32>
+// CHECK-NEXT:   func.return
+// CHECK-NEXT: }
+
+
+  func.func @choose_correct_tensor() {
+    %t0, %t1, %t2, %m0 = "test.op"() : () -> (tensor<8xf32>, tensor<8xf32>, tensor<8xf32>, memref<8xf32>)
+    %not_this = bufferization.to_tensor %m0 restrict : memref<8xf32>
+    %this = bufferization.to_tensor %m0 restrict writable : memref<8xf32>
+    %also_not_this = bufferization.to_tensor %m0 restrict : memref<8xf32>
```

Comment:
These seem not to be used?

Reply:
Well, `%this` is used, but what isn't used is `%not_this` and `%also_not_this` 🥹

Reply:
The outer two shouldn't be found by the discovery function because they're not exposing `writable` tensors.

Reply:
I see, I didn't spot it. It would probably be good to add the explanation in a comment somewhere, but that's more of a nit.

Reply:
As a meta point, I do feel that this might be another case of doing two things in one pass. Why not always create a tensor.empty, and then infer whether it should be reused in another pass? Also, isn't bufferization already supposed to do this sort of analysis?

Reply:
Thanks, added a comment as suggested.

I also thought that bufferization is supposed to do this analysis, but to keep the pass simple bufferization treats every `outs` as an init tensor, and with that as something having more semantic meaning than "shape-only" meaning of non-init-tensors. There is an `isInitTensor` function that I can't see being called in bufferization if I'm not mistaken, and there may be a simplifying assumption that treating everything as an init tensor means bufferization does not need to do such analysis, thus keeping it simpler. As any fix to upstream would take a little while as well as an mlir version update in xdsl, I'd recommend addressing it on our side.

Finally, the question is where to address it. Personally, I can easily fix it elsewhere, but I do like the idea of setting the args on translated ops in the best way we can.